lib/ABI/Analyses/ConvertFunctionsToCABI.cpp

```diff
@@ -254,7 +254,7 @@ class ConvertFunctionsToCABI {
     // And convert them.
     for (model::RawFunctionDefinition *Old : ToConvert) {
       auto &DT = llvm::cast<model::DefinedType>(*Model->makeType(Old->key()));
-      if (!checkVectorRegisterSupport(VectorVH, *Old)) {
+      if (Mode == "safe" && !checkVectorRegisterSupport(VectorVH, *Old)) {
```

Comment:
> we need _this_.

- This mode?

Because I'd rather we didn't touch user's function prototypes by default if we are very likely to break something, like in an untested case, but it's still useful to be able to force a conversion without recompiling revng when debugging/qa'ing stuff.

- Or this change?

Maybe I should have been more explicit in the code, but this change is literally "skip `checkVectorRegisterSupport` call if in 'unsafe' mode", as in, if you explicitly specify `--mode=unsafe` you most likely want the conversion attempt no matter what the prototype is like.

Reply:
1. The "safe" vs "unsafe" distinction should not exist in the first place. Let's not expand its usage.
   Let's drop this change.
2. I've taken a look at `checkVectorRegisterSupport`. It abuses `VerifyHelper` and it's prone to break whenever we make some significant change to the type system. This can be avoided by using `edges()`.
   1. Rename it `usesFloat`.
   2. Build the (inverse) graph of Types [as we're doing here](https://github.com/revng/revng/blob/develop/lib/Model/Processing.cpp#L36). This can be done once per `ConvertFunctionsToCABI` invocation.
   3. Use a visit from each leaf containing a float primitive to determine if they use floats.

   If you *really* don't like materializing the graph, you can use a recursive visit with `edges` but you need to ensure you don't end up in infinite recursion. In any case *do not* use `VerifyHelper`.

Reply:
1. 👌 

2.0. The thing is, I am pretty confident this predates `edges()` being a thing (but don't quote me on that, I haven't checked). And it's in maintenance mode - once we get proper float support, this doesn't need to exist anymore.

2.1. `usesFloat` - 👌

2.2. If we're doing this, we should probably do it properly, probably as a helper within model namespace (or at least, within abi::Definition) because I imagine this logic will be needed to implement _actual_ floating point support once we're at that point.

Either way, I'm hesitant to spend effort on these half-measures. We should either do it properly or don't touch it (tm) until we're ready to do it properly (in this case, when adding actual vector register support),

Reply:
Apparently it's younger, but maybe it was not well publicized.

```
commit 158272dee57ba4ca347412202567a30f4da85d5a
Author: Ivan Krysak <ivan@rev.ng>
Date:   Tue Feb 7 16:50:54 2023 +0200

    ABI: add an analysis for `ToCABI` conversion

commit b31e7b1f5e96b7f4400fb2eec0c7a379b18a4efd
Author: Alessandro Di Federico <ale@rev.ng>
Date:   Tue Jan 25 14:42:03 2022 +0100

    Introduce `model::Type::edges()`

    This makes building a graph for the type system very easy.
```

Anyway, OK, your observation makes sense. Hopefully we'll have floats soon.

Please add a comment not to just not use `VerifyHelper` but also about using `edges()`.